source/iterator/cdc.go

```diff
@@ -0,0 +1,411 @@
+		})
+	}
+
+	return shardGtids, nil
+}
+
+// constructRuleFilter constructs an SQL query for the binlogdata.Filter.Rules.
+func (c *cdc) constructRuleFilter(table, orderingColumn string, columns []string) (string, error) {
+	selectDataset := goqu.Dialect("mysql").Select()
+
+	if len(columns) > 0 {
+		keyColumnPresent := false
+		// add one to the capacity to have a space for the keyColumn
+		// if it's not present in the columns list.
+		cols := make([]any, 0, len(columns)+1)
+		for i := 0; i < len(columns); i++ {
+			if columns[i] == c.keyColumn {
+				keyColumnPresent = true
+			}
+
+			cols = append(cols, columns[i])
+		}
+
+		if !keyColumnPresent {
+			cols = append(cols, c.keyColumn)
+		}
+
+		selectDataset = selectDataset.Select(cols...)
+	}
+
+	query, _, err := selectDataset.
+		From(table).
+		Order(goqu.C(orderingColumn).Asc()).
+		ToSQL()
+	if err != nil {
+		return "", fmt.Errorf("construct rule filter query: %w", err)
+	}
+
+	return query, nil
+}
+
+// listen listens for VStream events.
+// If the VStream encountered an error the method will send it to the errCh channel.
+// All the data are sent to the records channel.
+func (c *cdc) listen(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+
+		default:
+			events, err := c.reader.Recv()
+			if err != nil {
+				c.errCh <- fmt.Errorf("read from vstream: %w", err)
+
+				return
+			}
+
+			var rowEvent *binlogdata.VEvent
+
+			for _, event := range events {
+				switch eventType := event.Type; {
+				case eventType == binlogdata.VEventType_VGTID && rowEvent != nil:
+					c.position.ShardGtids = event.Vgtid.ShardGtids
+
+					if err := c.processRowEvent(ctx, rowEvent); err != nil {
+						c.errCh <- fmt.Errorf("process row event: %w", err)
+
+						return
+					}
+
+				case eventType == binlogdata.VEventType_FIELD:
+					c.fields = event.FieldEvent.Fields
+
+				case eventType == binlogdata.VEventType_ROW:
+					if c.fields == nil {
+						// shouldn't happen cause VEventType_FIELD always comes before VEventType_ROW.
+						sdk.Logger(ctx).Warn().Msgf("i.fields is nil, skipping the row")
+
+						continue
+					}
+
+					rowEvent = event
+
+				default:
+				}
+			}
+		}
+	}
+}
+
+// processRowEvent makes rows from the event.RowEvent.RowChanges trusted,
+// constructs the resulting slice containing all needed sqltypes.Values,
+// transforms it to a sdk.Record and sends the record to a c.records channel.
+func (c *cdc) processRowEvent(ctx context.Context, event *binlogdata.VEvent) error {
+	for _, change := range event.RowEvent.RowChanges {
+		var (
+			valuesBefore []sqltypes.Value
+			valuesAfter  []sqltypes.Value
+			operation    = sdk.OperationCreate
+		)
+
+		switch after, before := change.After, change.Before; {
+		case after != nil && before != nil:
+			operation = sdk.OperationUpdate
+			valuesBefore = sqltypes.MakeRowTrusted(c.fields, before)
+			valuesAfter = sqltypes.MakeRowTrusted(c.fields, after)
+
+		case before != nil:
```

Comment:
Is it possible that `before == nil` and `after == nil` in a delete?

Reply:
The delete event always has the `before` field. The `before` field contains the values that were in the row before it was deleted with the database commit.